tensorflow/core/kernels/data/experimental/auto_shard_dataset_op_test.cc

```diff
@@ -0,0 +1,282 @@
+  Tensor step = CreateTensor<int64>(TensorShape({}),
+                                    {test_case.range_dataset_param.step});
+  Tensor range_dataset_tensor(DT_VARIANT, TensorShape({}));
+  TF_ASSERT_OK(MakeRangeDataset(start, stop, step, {DT_INT64},
+                                {TensorShape({})}, &range_dataset_tensor));
+
+  Tensor num_workers = test_case.num_workers;
+  Tensor index = test_case.index;
+  gtl::InlinedVector<TensorValue, 4> inputs({TensorValue(&range_dataset_tensor),
+                                             TensorValue(&num_workers),
+                                             TensorValue(&index)});
+  std::unique_ptr<OpKernelContext> auto_shard_dataset_context;
+  TF_ASSERT_OK(CreateAutoShardDatasetContext(
+      auto_shard_dataset_kernel.get(), &inputs, &auto_shard_dataset_context));
+
+  DatasetBase* auto_shard_dataset;
+  TF_ASSERT_OK(CreateDataset(auto_shard_dataset_kernel.get(),
+                             auto_shard_dataset_context.get(),
+                             &auto_shard_dataset));
+  core::ScopedUnref scoped_unref_auto_shard_dataset(auto_shard_dataset);
+
+  std::unique_ptr<IteratorContext> iterator_ctx;
+  TF_ASSERT_OK(
+      CreateIteratorContext(auto_shard_dataset_context.get(), &iterator_ctx));
+  std::unique_ptr<IteratorBase> iterator;
+  TF_ASSERT_OK(auto_shard_dataset->MakeIterator(iterator_ctx.get(),
+                                                kIteratorPrefix, &iterator));
+
+  bool end_of_sequence = false;
+  auto expected_outputs_it = test_case.expected_outputs.begin();
+  std::vector<Tensor> out_tensors;
+  while (!end_of_sequence) {
+    TF_EXPECT_OK(
+        iterator->GetNext(iterator_ctx.get(), &out_tensors, &end_of_sequence));
+    if (!end_of_sequence) {
+      EXPECT_LT(expected_outputs_it, test_case.expected_outputs.end());
+      TF_EXPECT_OK(ExpectEqual(out_tensors.back(), *expected_outputs_it));
+      expected_outputs_it++;
+    }
+  }
+  EXPECT_EQ(expected_outputs_it, test_case.expected_outputs.end());
+}
+
+INSTANTIATE_TEST_SUITE_P(AutoShardDatasetOpTest,
+                         ParameterizedAutoShardDatasetOpTest,
+                         ::testing::ValuesIn(std::vector<TestCase>(
+                             {TestCase1(), TestCase2(), TestCase3()})));
+
+TEST_F(AutoShardDatasetOpTest, InvalidArguments) {
+  int thread_num = 2, cpu_num = 2;
+  TF_ASSERT_OK(InitThreadPool(thread_num));
+  TF_ASSERT_OK(InitFunctionLibraryRuntime({}, cpu_num));
+
+  std::vector<TestCase> test_cases = {
+      IndexGreaterNumWorkersCase(), NegativeIndexTestCase(),
+      NegativeNumWorkersTestCase(), ZeroNumWorkersTestCase()};
+  for (const auto& test_case : test_cases) {
+    std::unique_ptr<OpKernel> auto_shard_dataset_kernel;
+    TF_ASSERT_OK(CreateAutoShardDatasetOpKernel(
+        test_case.expected_output_dtypes, test_case.expected_output_shapes,
+        &auto_shard_dataset_kernel));
+
+    Tensor start = CreateTensor<int64>(TensorShape({}),
```

Comment:
Some repeated boilerplate code between here and the `GetNext` test harness above. See if it is possible to collapse the initialization stuff into one function?

Reply:
@frankchn Yes, there are some repeated codes in different API tests, such as `CreateDataset`, `MakeIterator`, `CreateIteratorContext`, and `MakeIterator`. We tried to wrap them as a function inside the test class (e.g. `AutoShardDatasetOpTest`) before, but it makes users difficult to understand the test code and not easy to change the code for each individual test if needed in the future (cc: @jsimsa). 

Here is some code we tried before:
```C++
class AutoShardDatasetOpTest : public DatasetOpsTestBase {
 protected:
  Status CreateAutoShardDatasetOpKernel( ... );

  Status CreateAutoShardDatasetContext( ... );

  Status CreateAutoShardDataset( ... );
  
  Status CreateAutoShardIteratorContext( ... );

  Status CreateAutoShardIteratorContext( ... );

 private:
   std::unique_ptr<OpKernel> auto_shard_dataset_kernel_;
   std::unique_ptr<OpKernelContext> auto_shard_dataset_context_;
   DatasetBase* auto_shard_dataset_;
   std::unique_ptr<IteratorContext> iterator_ctx_;
   std::unique_ptr<IteratorContext> iterator_ctx_;  
};
```